packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -935,6 +933,53 @@ ArbCore::reorgToFirstMatchingMachineCheckpoint(
     return rocksdb::Status::NotFound();
 }
 
+std::unique_ptr<MachineThread> ArbCore::getMachineThreadFromCheckpoint(
+    const CheckpointVariant& current_database_checkpoint,
+    const std::function<bool(const MachineOutput&)>& check_output,
+    ReadWriteTransaction& tx,
+    ValueCache& value_cache) {
+    if (!std::holds_alternative<MachineStateKeys>(
+            current_database_checkpoint)) {
+        // Just holds MachineOutput, so continue until full checkpoint found
```

Comment:
Is it correct to not check the cache in this case? Eg if the latest valid machine in the cache is after this checkpoint, we can just use it. I think this can be fixed by just moving this to right above the `try` since everything before just needs a `MachineOutput`

Reply:
In `findFirstMatching`, we don't know what the target gas is, so there is no way to know if the item found in cache is within `database_load_gas_cost` of the target.  Because of this, a cache entry that is `database_load_gas_cost * 4` away from the target, and if we kept looking through database until we got to a full checkpoint, it is possible that checkpoint would be less than `database_load_gas_cost`.